### PR TITLE
[gitea] improve security and usabillity

### PR DESCRIPTION
**Pull request by FrankZisko**
[gitea] improve security: We verify the download using the authors pgp key. Instead of expose the web service to the world to finish the installation, we lock the service, generate the admin user and write a modified config file.

### CODE REVIEW

**Comment by FrankZisko** [comment]
Oh, sure directly after requesting: I've seen, some usernames aren't replaced with isabell (test0r) and some are wrong (isabelle). How to change afterwards?

File: source/guide_gitea.rst [hunk 1]
```diff
@@ -29,62 +28,124 @@ of the then-popular Gogs_, but quickly became more active and bigger than the or
   * :manual:`supervisord <daemons-supervisord>`
   * :manual:`domains <web-domains>`
 
+
 Prerequisites
 =============
 
-.. include:: includes/my-print-defaults.rst
```
Comment (luto) [inline source/guide_gitea.rst:35]: This snippet is included in all guides, which use the mysql server so readers know where to get the mysql PW. Please use it instead of the more pipey-version below. If you'd like to change this, please open an issue so we can change the snippet and improve all mysql dependent guides that way.

File: source/guide_gitea.rst [hunk 2]
```diff
@@ -29,62 +28,124 @@ of the then-popular Gogs_, but quickly became more active and bigger than the or
...
-  [isabell@stardust ~]$
+  [isabell@stardust ~]$ my_print_defaults --defaults-file=~/.my.cnf client | grep -- --password | awk --field-separator = '{ print $2 }'
+  xG9MD-15A4aMGsln,MTr
+  [isabell@stardust ~]$ 
+
+... and some random alpha-num chars for the security key. For this we pipe some random characters from the ``/dev/urandom`` device into the translate tool ``tr`` to delete the complementary of alpha-numeric characters until we have 32 characters.
+
+.. code-block:: console
+
+  [isabell@stardust ~]$ cat /dev/urandom | tr --delete --complement 'a-zA-Z0-9' | fold --width=32 | head --lines=1
```
Comment (luto) [inline source/guide_gitea.rst:137]: newer versions of U7 offer "pwgen", which enables your to shorten this line to `pwgen 32`.

File: source/guide_gitea.rst [hunk 3]
```diff
@@ -96,93 +157,276 @@ and a new config file at ``~/gitea/custom/conf/app.ini`` to specify the port and
...
+.. note:: The minimum improvements are written in the sections of ``[database]``, ``[security]`` and ``[service]``.
 
-.. note::
+.. warning:: Replace ``isabell`` with your username, fill the database password ``PASSWD =`` with yours and enter the generated random into ``SECRET_KEY =``.
 
-    Gitea is running on port 9000.
+.. code-block:: ini
 
-.. include:: includes/web-backend.rst
+  APP_NAME = Gitea
```
Comment (luto) [inline source/guide_gitea.rst:176]: Like with the previous PR, this is the full configuration including all default values (e.g. `OFFLINE_MODE`, `PASSWORD_HASH_ALGO` and many more). Please remove them and only include configuration that actually needs changing. Having the whole default config in where makes it hard to update the guide in the future.

File: source/guide_gitea.rst [hunk 4]
```diff
@@ -96,93 +157,276 @@ and a new config file at ``~/gitea/custom/conf/app.ini`` to specify the port and
...
+  RUN_USER = isabell
+  RUN_MODE = prod ; Either "dev", "prod" or "test", default is "dev".
+  
+  [server]
+  HTTP_PORT            = 9000
```
Comment (luto) [inline source/guide_gitea.rst:181]: Feel free to omit this and switch to the default port 9000 to save a line of config.

File: source/guide_gitea.rst [hunk 5]
```diff
@@ -96,93 +157,276 @@ and a new config file at ``~/gitea/custom/conf/app.ini`` to specify the port and
...
+  SHOW_MILESTONES_DASHBOARD_PAGE         = true
+  DEFAULT_ALLOW_CREATE_ORGANIZATION      = true
+  DEFAULT_ORG_VISIBILITY                 = private ; [public, limited, private]
+  DEFAULT_ORG_MEMBER_VISIBLE             = false
+  DEFAULT_ENABLE_TIMETRACKING            = true
+  DEFAULT_KEEP_EMAIL_PRIVATE             = true
+  NO_REPLY_ADDRESS                       = noreply.isabell.uber.space
+  
+  [repository]
+  ROOT = /home/isabell/gitea/repositories
```
Comment (luto) [inline source/guide_gitea.rst:233]: The default path ` ~/gitea-repositories` seems fine to me and makes the new guide more compatible to installations based on the old one.

File: source/guide_gitea.rst [hunk 6]
```diff
@@ -96,93 +157,276 @@ and a new config file at ``~/gitea/custom/conf/app.ini`` to specify the port and
...
+  [isabell@stardust ~]$ gem install asciidoctor
+  Fetching asciidoctor-2.0.10.gem
+  WARNING:  You don't have /home/isabell/.gem/ruby/2.7.0/bin in your PATH,
+  	  gem executables will not run.
+  Successfully installed asciidoctor-2.0.10
+  1 gem installed
+  [isabell@stardust ~]$
 
-Create ``~/etc/services.d/gitea.ini`` with the following content:
+.. note:: Don't be irritated by the warning that the bin folder isn't in the path. Uberspace is taking care of it and provide it via ``/opt/uberspace/etc/isabell/binpaths/ruby/asciidoctor``. You can check with ``[isabell@stardust ~]$ which asciidoctor``.
```
Comment (luto) [inline source/guide_gitea.rst:275]: `/opt/uberspace/etc/isabell/binpaths` is not documented publicly. While we don't consider this a secret (it's right there in front of you ;)), we try to support details, that are documented publicly, as long as possible. If not necessary we try to not include too many implementation details outside of the "background" section in our manual.

Please shorten this to "Uberspace is taking care of it. You can check with (...)".

File: source/guide_gitea.rst [hunk 7]
```diff
@@ -96,93 +157,276 @@ and a new config file at ``~/gitea/custom/conf/app.ini`` to specify the port and
...
+
+Gitea admin user
+----------------
+
+We generate a safe password for the admin user. For this we use ``head`` to pipe some random characters from the ``/dev/urandom`` device into the translate tool ``tr`` to delete the complementary of alpha-numeric characters until we have 16 characters and then write into a variable to use it in the next two steps.
+
+.. code-block:: console
 
-Point your browser to the ``/install`` path on your domain to finish the installation.
+  [isabell@stardust ~]$ ADMPWD=$(head /dev/urandom | tr --complement --delete A-Za-z0-9 | head --bytes=16 ; echo '')
```
Comment (luto) [inline source/guide_gitea.rst:313]: see above, please use `pwgen 16`.

File: source/guide_gitea.rst [hunk 8]
```diff
@@ -96,93 +157,276 @@ and a new config file at ``~/gitea/custom/conf/app.ini`` to specify the port and
...
+.. include:: includes/supervisord.rst
+
+.. note:: The status of gitea must be ``RUNNING``. If not, check the configuration file ``~/gitea/custom/conf/app.ini``.
+
+Uberspace web backend
+---------------------
+
+.. code-block:: console
+
+  [isabell@stardust ~]$ uberspace web backend set / --http --port 9000
```
Comment (luto) [inline source/guide_gitea.rst:387]: Please us the web backend snippet here:

```
.. note::

    gittea is running on port 9000.

.. include:: includes/web-backend.rst
```

I agree that it leads to a less pretty result, but makes the guide much more manageable for us.

File: source/guide_gitea.rst (hunk 1 above)
Comment (FrankZisko) [reply source/guide_gitea.rst:35]: Done

File: source/guide_gitea.rst (hunk 2 above)
Comment (FrankZisko) [reply source/guide_gitea.rst:137]: Done.

File: source/guide_gitea.rst (hunk 4 above)
Comment (FrankZisko) [reply source/guide_gitea.rst:181]: It's already port 9000. I'm sorry, I don't see it.

File: source/guide_gitea.rst (hunk 3 above)
Comment (FrankZisko) [reply source/guide_gitea.rst:176]: Hopefully it's slim enough now. :-) 

I think some security improvements (e.g. to allow lower case passwords only as referred to xkcd) are ok. Email confirmation depends to configure [mailer] section.

File: source/guide_gitea.rst (hunk 5 above)
Comment (FrankZisko) [reply source/guide_gitea.rst:233]: Done.

File: source/guide_gitea.rst (hunk 6 above)
Comment (FrankZisko) [reply source/guide_gitea.rst:275]: Done.

File: source/guide_gitea.rst (hunk 7 above)
Comment (FrankZisko) [reply source/guide_gitea.rst:313]: Done.

File: source/guide_gitea.rst (hunk 8 above)
Comment (FrankZisko) [reply source/guide_gitea.rst:387]: Done.

File: source/guide_gitea.rst (hunk 4 above)
Comment (luto) [reply source/guide_gitea.rst:179]: yes, 9000 is the default so you can just omit that line entirely.

File: source/guide_gitea.rst (hunk 3 above)
Comment (luto) [reply source/guide_gitea.rst:174]: yes, any security improvements are certainly welcome and should stay! The lines that just set the default values are the ones that bothered me :sweat_smile:

File: source/guide_gitea.rst (hunk 4 above)
Comment (FrankZisko) [reply source/guide_gitea.rst:179]: I copied it from the last version of this *.ini.

```
[server]
HTTP_PORT = 9000
DOMAIN = isabell.uber.space
ROOT_URL = https://%(DOMAIN)s

[service]
DISABLE_REGISTRATION = true
```

Comment (FrankZisko) [reply source/guide_gitea.rst:179]: They changed the default port to 3000. Should we change this to, or stay at 9000?

-- https://docs.gitea.io/en-us/config-cheat-sheet/

File: source/guide_gitea.rst (hunk 3 above)
Comment (FrankZisko) [reply source/guide_gitea.rst:174]: I checked again. 

**`[server]`**
I removed the `LFS` options complete. The default paths are looking good, and the security key will be created in on of the next steps, so we can remove this here.

I think `OFFLINE_MODE = true` is an important privacy option, so there will be no traffic to gravatar and others.

**`[service]`** and **`[mailer]`**
I think the options written there a a combination of security, so the platform is not open for any others by default and keeping privacy of new users by default.

File: source/guide_gitea.rst (hunk 4 above)
Comment (luto) [reply source/guide_gitea.rst:179]: 3000 sounds good!

**Comment by FrankZisko** [comment]
What is the review status. Can I do anything?

**Comment by luto** [comment]
Like the other bigger reviews, this is on hold since we're quite busy with user support at the moment, sorry.

**Comment by FrankZisko** [comment]
No problem. Thx for this information.

**Comment by luto** [comment]
Thank you very much for all your work! merged.

**Comment by Schlumpf** [comment]
Merged?
I am confused, Github says "closed"

**Comment by luto** [comment]
I merged it on the command line, not using the interface. Have a look: https://lab.uberspace.de/guide_gitea.html

**Comment by Schlumpf** [comment]
OK, Sorry for the noise
